indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,197 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
-      // - header 1: in source tree (has canonical path), uses namespace {..}
-      // - header 2: in source tree (has canonical path), uses namespace {..}
-      // - generated C++ file: in build tree only (no canonical path), and
-      //   includes header 1 and 2.
-      // We will not emit a symbol that connects the anonymous namespace
-      // in header 1 and header 2. However, that is OK, because it is unclear
-      // how to handle this case anyways, and anonymous namespaces are
-      // rarely (if ever) used in headers.
-      return {};
-    }
-    fmt::format_to(std::back_inserter(this->scratchBuffer), "$ANON/{}\0",
-                   path->asStringView());
-  } else {
-    llvm::raw_string_ostream os(this->scratchBuffer);
-    namespaceDecl->printQualifiedName(os);
-    absl::StrReplaceAll({{"::", "/"}}, &this->scratchBuffer);
+  auto symbol = getSymbol();
+  if (symbol.empty()) {
+    return "";
   }
-  std::string out{
-      fmt::format("c . todo-pkg todo-version {}/", this->scratchBuffer)};
-  this->scratchBuffer.clear();
   auto [newIt, inserted] =
-      this->declBasedCache.insert({namespaceDecl, std::move(out)});
+      this->declBasedCache.insert({decl, std::move(symbol)});
   ENFORCE(inserted);
   return std::string_view(newIt->second);
 }
 
+std::optional<std::string_view>
+SymbolFormatter::getContextSymbol(const clang::DeclContext *declContext) {
+  if (auto namespaceDecl = llvm::dyn_cast<clang::NamespaceDecl>(declContext)) {
+    return this->getNamespaceSymbol(namespaceDecl).value_or("");
+  }
+  if (auto tagDecl = llvm::dyn_cast<clang::TagDecl>(declContext)) {
+    return this->getTagSymbol(tagDecl);
+  }
+  if (llvm::isa<clang::TranslationUnitDecl>(declContext)
+      || llvm::isa<clang::ExternCContextDecl>(declContext)) {
+    auto decl = llvm::dyn_cast<clang::Decl>(declContext);
+    return this->getSymbolCached(
+        decl, [&]() -> std::string { return "c . todo-pkg todo-version "; });
+  }
+  // TODO: Handle all cases of DeclContext here:
+  // Done
+  // - TranslationUnitDecl
+  // - ExternCContext
+  // - NamespaceDecl
+  // - TagDecl
+  // Pending:
+  // - OMPDeclareReductionDecl
+  // - OMPDeclareMapperDecl
+  // - FunctionDecl
+  // - ObjCMethodDecl
+  // - ObjCContainerDecl
+  // - LinkageSpecDecl
+  // - ExportDecl
+  // - BlockDecl
+  // - CapturedDecl
+  return std::nullopt;
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getTagSymbol(const clang::TagDecl *tagDecl) {
+  return this->getSymbolCached(tagDecl, [&]() -> std::string {
+    auto optContextSymbol = this->getContextSymbol(tagDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    auto contextSymbol = optContextSymbol.value();
+    if (!tagDecl->getDeclName().isEmpty()) {
+      llvm::raw_string_ostream os(this->scratchBuffer);
+      static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
+      std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
```

Comment:
For one, it'll make the code longer. Compare:

```
// With stream API
os << contextSymbol;
static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
os << '#';
std::string out{os.str()};
this->scratchBuffer.clear();
return out;

// Using fmt::format
static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
this->scratchBuffer.clear();
return out;
```

In either case, we still need to perform the copy out of the scratch buffer because this function needs to return an owned string. Did you have any benefits in mind for using the stream API?

Reply:
If you do this a lot, you could use some helper functions/operators/RAII make this roughly

```
ClearScratch clear(this);
os << contextSymbol << mumble(tagDecl) << '#';
return {os.str()};
```

...you can probably think of better names and how to make it even neater. For example, if you need to hand over a string, it begs the question why use the scratch buffer at all.

The important part from my point of view is it concentrates the string format into one line, and that's the important part of what's happening. Not a big deal, but if you're going to do it a lot, the burden on the reader can add up over time.

Reply:
> If you do this a lot, you could use some helper functions/operators/RAII make this roughly

This is mostly very low-level code, and if it goes the way of scip-ruby, the formatting specifics will likely end up being gathered into one central function instead of happening in several places. https://github.com/sourcegraph/scip-ruby/blob/scip-ruby/master/scip_indexer/SCIPUtils.cc#L43

Right now, there is a bit of duplication, because I'm waiting to have more code paths to figure out what the common pattern is, before trying to factor it out.